php_thread.go

```diff
@@ -46,5 +99,55 @@ func (thread *phpThread) pinString(s string) *C.char {
 
 // C strings must be null-terminated
 func (thread *phpThread) pinCString(s string) *C.char {
-	return thread.pinString(s+"\x00")
+	return thread.pinString(s + "\x00")
+}
+
+//export go_frankenphp_before_script_execution
+func go_frankenphp_before_script_execution(threadIndex C.uintptr_t) *C.char {
+	// returning nil signals the thread to stop
+	if threadsAreDone.Load() {
+		return nil
+	}
+	thread := phpThreads[threadIndex]
+
+	// if the thread is not ready, set it up
+	if !thread.isReady.Load() {
+		thread.isReady.Store(true)
+		thread.done = make(chan struct{})
+		if thread.onStartup != nil {
+			thread.onStartup(thread)
+		}
+		if threadsAreBooting.Load() {
+			threadsReadyWG.Done()
+			threadsReadyWG.Wait()
+		}
```

Comment:
This is a perfect case of where locks are probably better, so we can block until the lock is released and threads are done booting. Otherwise, the race condition is that if this function is called while booting, the thread will never be ready.

Can be something like:

```suggestion
		threadsAreBooting.RLock()
		threadsReadyWG.Done()
		threadsReadyWG.Wait()
		threadsAreBooting.RUnlock()
```

abusing rwlock so when booting, we take an exclusive lock, but here we take a shared lock. Thus if we are still booting, we will wait, but once booted, we can just fly right through here with no contention.

Reply:
Using Locks here will lead to a different behavior.
The `Waitgroup` will actually wait for all threads to reach this place before `initPHPThreads` returns. Doing this with Locks will make `initPHPThreads` return immediately and cause a panic.

Reply:
The reason why we have to wait here is to make sure that all threads have called `(void)ts_resource(0)` before any worker scripts are executed. PHP would sometimes segfault if script execution happens before all threads have fetched their initial resources.